src/server/src/main/java/com/spotify/reaper/service/SegmentRunner.java

```diff
@@ -613,29 +629,47 @@ public void handle(int repairNumber, Optional<ActiveRepairService.Status> status
             break;
         }
       }
-   // New repair API
+      // New repair API – Cassandra-2.2 onwards
       if(progress.isPresent()) {
         switch (progress.get()) {
           case START:
-            renewLead();
-            DateTime now = DateTime.now();
-            context.storage.updateRepairSegment(currentSegment.with()
-                .state(RepairSegment.State.RUNNING)
-                .startTime(now)
-                .build(segmentId));
-            LOG.debug("updated segment {} with state {}", segmentId, RepairSegment.State.RUNNING);
+            try {
+              if (renewLead()) {
+                DateTime now = DateTime.now();
+
+                context.storage.updateRepairSegment(currentSegment.with()
+                    .state(RepairSegment.State.RUNNING)
+                    .startTime(now)
+                    .build(segmentId));
+
+                LOG.debug("updated segment {} with state {}", segmentId, RepairSegment.State.RUNNING);
+                break;
+              }
+            } catch (AssertionError er) {
+              // ignore. segment repair has since timed out.
+            }
+            segmentFailed.set(true);
             break;
 
           case SUCCESS:
-            LOG.debug(
-                    "repair session succeeded for segment with id '{}' and repair number '{}'",
+            try {
+              if (segmentFailed.get()) {
+                LOG.debug(
+                    "Got SUCCESS for segment with id '{}' and repair number '{}', but it had already timed out",
                     segmentId, repairNumber);
+              } else if (renewLead()) {
```

Comment:
I have some concerns that we may be issuing a lot of LWTs which could increase the risk of timeout, and subsequently give the impression to Reaper that another instance has the lead (while it has not).
Instead of calling `renewLead()` so many times, do you think we could instead use a `hasLead()` method that performs a standard QUORUM read that can be retried at LOCAL_QUORUM in case of timeout ?

Reply:
how many `SUCCESS` jmx notifications are we expecting?

Reply:
One per segment.

Reply:
Isn't this minor? Don't we call this anyway every 30 seconds on each segment?

The existing `hasLeadOnSegment(..)` method is just a fail-safe wrapper to `renewLead(..)`.
My gut feeling here is this isn't worth adding the code complexity for any new non-LWT `hasLead(..)` method.

We also have the metrics in place to be evaluating this onwards.
Can we come back to it if it's a problem, with metrics to give us better insight?

Reply:
Sure, let's do this.